red53_agent.go

```diff
@@ -0,0 +1,203 @@
+
+var DNSName = "servicediscovery.local"
+
+func logErrorAndFail(err error) {
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func logErrorNoFatal(err error) {
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+type config struct {
+	HostedZoneId string
+	RedisHost    string
+}
+
+var configuration config
+
+func getDNSHostedZoneId() (string, error) {
+	r53 := route53.New(session.New())
+	params := &route53.ListHostedZonesByNameInput{
+		DNSName: aws.String(DNSName),
+	}
+
+	zones, err := r53.ListHostedZonesByName(params)
+
+	if err == nil && len(zones.HostedZones) > 0 {
+		return aws.StringValue(zones.HostedZones[0].Id), nil
+	}
+
+	return "", err
+}
+
+// Modify (or create) the A record for this serviceName, adding the private IP of the host.
+func modifyDNSRecord(serviceName string, ips []string) error {
+	var err error
+	r53 := route53.New(session.New())
+
+	aValues := make([]*route53.ResourceRecord, 0)
+	// Put all the IPs in one A record
+	for _, ip := range ips {
+		aValues = append(aValues, &route53.ResourceRecord{Value: aws.String(ip)})
+	}
+
+	// This API call creates a new DNS record for this service
+	params := &route53.ChangeResourceRecordSetsInput{
+		ChangeBatch: &route53.ChangeBatch{
+			Changes: []*route53.Change{
+				{
+					Action: aws.String(route53.ChangeActionUpsert),
+					ResourceRecordSet: &route53.ResourceRecordSet{
+						Name: aws.String(serviceName + "." + DNSName),
+						// It creates an A record with the name of the service
+						Type:            aws.String(route53.RRTypeA),
+						ResourceRecords: aValues,
+						TTL:             aws.Int64(defaultTTL),
+					},
+				},
+			},
+			Comment: aws.String("Service Discovery Created Record"),
+		},
+		HostedZoneId: aws.String(configuration.HostedZoneId),
+	}
+	_, err = r53.ChangeResourceRecordSets(params)
+	logErrorNoFatal(err)
+	log.Debug("Record " + serviceName + "." + DNSName + " updated.")
+	return err
+}
+
+var redisClient *redis.Client
+
+func processServicePings(servicePings []string) {
+	var serviceMap = make(map[string][]string)
+
+	for _, servicePing := range servicePings {
+		pingChunks := strings.Split(servicePing, "_")
+		serviceName := pingChunks[0]
+		serviceIp := pingChunks[1]
+		serviceMap[serviceName] = append(serviceMap[serviceName], serviceIp)
+	}
+
+	for serviceName, serviceIps := range serviceMap {
+		modifyDNSRecord(serviceName, serviceIps)
+	}
+}
+
+func fetchActiveServices() {
+	var servicePings []string
+	var sum int
+	var err error
+	log.Debug("Fetching active services ...")
+
+	// Fetch last "fetchLast" seconds of service pings
+	now := time.Now().Unix()
+	epoch := now - fetchLast // seconds
+
+	// Try several times with exponential backoff.
+	sum = 1
+	for {
+		servicePings, err = redisClient.ZRange("redecs:service_pings", epoch, now).Result()
+
+		if err == nil {
+			break
+		}
+		if sum > 8 {
+			// Bail out if this is failing.
+			logErrorAndFail(err)
+		}
+		time.Sleep(time.Duration(sum) * time.Second)
```

Comment:
Confusing cast: time.Duration(sum) isn't a meaningful duration. sum * time.Second should already be a duration?

time.Duration(sum) would technically be 5 nanoseconds?